src/mmap_unix.rs

```diff
@@ -0,0 +1,161 @@
+
+use std::os::unix::io::AsRawFd;
+
+/// A backend driver to access guest's physical memory by mmapping guest's memory into the current
+/// process.
+/// For a combination of 32-bit hypervisor and 64-bit virtual machine, only partial of guest's
+/// physical memory may be mapped into current process due to limited process virtual address
+/// space size.
+#[derive(Debug)]
+pub struct MmapRegion {
+    addr: *mut u8,
+    size: usize,
+}
+
+// Send and Sync aren't automatically inherited for the raw address pointer.
+// Accessing that pointer is only done through the stateless interface which
+// allows the object to be shared by multiple threads without a decrease in
+// safety.
+unsafe impl Send for MmapRegion {}
+unsafe impl Sync for MmapRegion {}
+
+impl MmapRegion {
+    /// Creates an anonymous shared mapping of `size` bytes.
+    ///
+    /// # Arguments
+    /// * `size` - Size of memory region in bytes.
+    pub fn new(size: usize) -> io::Result<Self> {
+        // This is safe because we are creating an anonymous mapping in a place not already used by
+        // any other area in this process.
+        let addr = unsafe {
+            libc::mmap(
+                null_mut(),
+                size,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_ANONYMOUS | libc::MAP_SHARED | libc::MAP_NORESERVE,
+                -1,
+                0,
+            )
+        };
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+        Ok(Self {
+            addr: addr as *mut u8,
+            size,
+        })
+    }
+
+    /// Maps the `size` bytes starting at `offset` bytes of the given `fd`.
+    ///
+    /// # Arguments
+    /// * `fd` - File descriptor to mmap from.
+    /// * `size` - Size of memory region in bytes.
+    /// * `offset` - Offset in bytes from the beginning of `fd` to start the mmap.
+    pub fn from_fd(fd: &AsRawFd, size: usize, offset: libc::off_t) -> io::Result<Self> {
+        // This is safe because we are creating a mapping in a place not already used by any other
+        // area in this process.
+        let addr = unsafe {
+            libc::mmap(
+                null_mut(),
+                size,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset,
+            )
+        };
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+        Ok(Self {
+            addr: addr as *mut u8,
+            size,
+        })
+    }
+
+    /// Returns a pointer to the beginning of the memory region.  Should only be
+    /// used for passing this region to ioctls for setting guest memory.
+    pub fn as_ptr(&self) -> *mut u8 {
+        self.addr
+    }
+}
+
+pub trait AsSlice {
+    unsafe fn as_slice(&self) -> &[u8];
+
+    #[allow(clippy::mut_from_ref)]
+    unsafe fn as_mut_slice(&self) -> &mut [u8];
+}
```

Comment:
How about moving the AsSlice trait into file mmap.rs? So mmap.rs file defines and makes use of the AsSlice trait, and mmap_unix.rs/mmap_window.rs implements AsSlice for MmapRegion.
diff --git a/src/mmap.rs b/src/mmap.rs
index f5cee36..9f40c8e 100644
--- a/src/mmap.rs
+++ b/src/mmap.rs
@@ -29,16 +29,19 @@ use guest_memory::*;
 use volatile_memory::VolatileMemory;
 use Bytes;
 
-#[cfg(unix)]
-use mmap_unix::AsSlice;
 #[cfg(unix)]
 pub use mmap_unix::MmapRegion;
 
-#[cfg(windows)]
-use mmap_windows::AsSlice;
 #[cfg(windows)]
 pub use mmap_windows::MmapRegion;
 
+pub(crate) trait AsSlice {
+    unsafe fn as_slice(&self) -> &[u8];
+
+    #[allow(clippy::mut_from_ref)]
+    unsafe fn as_mut_slice(&self) -> &mut [u8];
+}
+
 /// Errors that can happen when creating a memory map
 #[derive(Debug)]
 pub enum MmapError {

Reply:
Done